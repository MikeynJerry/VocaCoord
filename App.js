import AppNavigator from './src/AppNavigator';
export default AppNavigator;