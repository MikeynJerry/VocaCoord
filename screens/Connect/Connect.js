import React, { Component } from 'react';
import { View, Text, ActivityIndicator } from 'react-native';
import { Button, TextInput, HelperText } from 'react-native-paper';
import { Divider } from 'react-native-elements';
import ClusterWS from 'clusterws-client-js';
import { classExists, getWords } from '../../firebase';
import {
  errors,
  CODE_LENGTH,
  NO_CLASS,
  NO_WORDS
} from '../../constants/Errors';

let wentBack = false;
export default class ConnectScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      classCode: '',
      isConnecting: false,
      classCodeError: false,
      errorCode: ''
    };
  }

  componentDidMount() {
    wentBack = false;
  }

  componentWillUnmount() {
    wentBack = true;
    if (this.socket && this.socket.disconnect) this.socket.disconnect();
  }

  isConnected = () => this.setState({ isConnecting: false });

  handleClassCodeChange = classCode =>
    this.setState({ classCode }, () => {
      const { classCodeError } = this.state;
      if (classCodeError) this.validateCodeLength();
    });

  validateClass = () => {
    const { classCode } = this.state;

    if (!this.validateCodeLength()) return;
    this.setState({ isConnecting: true, classCodeError: false, errorCode: '' });
    classExists(classCode).then(classroom => {
      if (!classroom)
        return this.setState({
          isConnecting: false,
          classCodeError: true,
          errorCode: NO_CLASS
        });
      return getWords(classroom).then(words => {
        if (words.length === 0)
          return this.setState({
            isConnecting: false,
            classCodeError: true,
            errorCode: NO_WORDS
          });
        return this.connectToClass(words);
      });
    });
  };

  connectToClass = words => {
    const { classCode } = this.state;
    this.socket = new ClusterWS({
      url: 'wss://temp-vocacoord.herokuapp.com/'
    });
    this.socket.on('connect', () => {
      this.channel = this.socket.subscribe(classCode);
      const {
        navigation: { navigate }
      } = this.props;
      if (wentBack) return;
      navigate('ClassroomScreen', {
        channel: this.channel,
        callback: this.isConnected,
        words
      });
    });
    /*this.socket.on('error', err => {
      console.error('error: ', err); // <--- commented out because it was throwing an error
    }); */
    //this.socket.on('disconnect', () => this.channel.unsubscribe()); <--- commented out because it broke
  }

  validateCodeLength = () => {
    const { classCode } = this.state;
    let { errorCode } = this.state;
    let classCodeError = false;
    if (!classCode || classCode.length !== 4) {
      classCodeError = true;
      errorCode = CODE_LENGTH;
    }
    this.setState({ classCodeError, errorCode });
    return !classCodeError;
  };

  render() {
    const { styles } = this.props;
    const { isConnecting, classCode, classCodeError, errorCode } = this.state;
    return (
      <View style={styles.container}>
        <View>
          <TextInput
            onBlur={this.validateCodeLength}
            style={styles.textbox}
            label="Class Code"
            mode="outlined"
            value={classCode}
            onChangeText={this.handleClassCodeChange}
            error={classCodeError}
            autoCapitalize="characters"
          />
          <HelperText type="error" visible={classCodeError}>
            {errors[errorCode]}
          </HelperText>
          <Divider style={styles.divider} />
          {isConnecting ? (
            <ActivityIndicator size="large" color="#ffa500" />
          ) : (
            <Button
              color="#ffa500"
              style={styles.button}
              mode="contained"
              onPress={() => this.validateClass()}
            >
              <Text style={styles.buttonText}>Connect</Text>
            </Button>
          )}
        </View>
      </View>
    );
  }
}
