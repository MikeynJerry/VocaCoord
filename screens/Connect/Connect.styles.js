import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center"
  },
  button: {
    minWidth: "60%",
    maxWidth: "60%"
  },
  textbox: {
    minWidth: "60%",
    maxWidth: "60%"
  },
  divider: {
    height: "10%",
    backgroundColor: "#fff"
  },
  buttonText: {
    fontSize: 24,
    fontWeight: "bold",
    color: "black"
  }
});
