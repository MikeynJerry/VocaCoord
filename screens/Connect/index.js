import Connect from "./Connect";
import enhance from "./Connect.enhancer";

export default enhance(Connect);
