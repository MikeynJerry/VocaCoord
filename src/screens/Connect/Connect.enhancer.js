import { withStyles } from "../../utility/withStyles";
import { styles } from "./Connect.styles";

export default withStyles(styles);
