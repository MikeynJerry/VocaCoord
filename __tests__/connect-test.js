import 'react-native';
import React from 'react';
import App from '../App';
import renderer from 'react-test-renderer';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';
import Connect from '../screens/Connect/Connect';
import { StyleSheet } from "react-native";
import {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { configure } from 'rsvp';

configure({ adapter: new Adapter()});

describe('Addition', () => {
    it('knows that 2 and 2 make 4', () => {
      expect(2 + 2).toBe(4);
    });
  });

  describe('Connect', () => {
    it('Should render Connect Screen component', () => {
        const styles = StyleSheet.create({
            container: {
              flex: 1,
              backgroundColor: "#fff",
              alignItems: "center",
              justifyContent: "center"
            },
            button: {
              minWidth: "60%",
              maxWidth: "60%"
            },
            textbox: {
              minWidth: "60%",
              maxWidth: "60%"
            },
            divider: {
              height: "10%",
              backgroundColor: "#fff"
            },
            buttonText: {
              fontSize: 24,
              fontWeight: "bold",
              color: "black"
            }
          });
        
        const tree = renderer
        .create(<Connect styles={styles} />)
        .toJSON();
        expect(tree).toMatchSnapshot();
    });
  }); 

  
  describe('ValidateClassCode', () => {
      it('Should throw error for bad class code', () => {
        
        const wrapper = shallow(<Connect />);
        const { validateClassCode } = wrapper.dive().instance();
        expect(validateClassCode().toBe(false));
      });
  });      