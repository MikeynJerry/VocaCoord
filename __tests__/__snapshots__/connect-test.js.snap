// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connect Should render Connect Screen component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <View
      style={
        Array [
          undefined,
          Object {
            "maxWidth": "60%",
            "minWidth": "60%",
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 6,
            },
            Object {
              "borderColor": "rgba(0, 0, 0, 0.54)",
              "borderRadius": 4,
              "borderWidth": 1,
            },
          ]
        }
      />
      <Text
        numberOfLines={1}
        pointerEvents="none"
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
              "textAlign": "left",
              "writingDirection": "ltr",
            },
            Object {
              "backgroundColor": "#f6f6f6",
              "color": "transparent",
              "fontFamily": "Helvetica Neue",
              "fontSize": 12,
              "left": 8,
              "opacity": 0,
              "paddingHorizontal": 4,
              "position": "absolute",
              "top": 0,
              "transform": Array [
                Object {
                  "scaleX": 0,
                },
              ],
            },
          ]
        }
      >
        Class Code
      </Text>
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          onLayout={[Function]}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Helvetica Neue",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Object {
                "color": "#6200ee",
                "fontFamily": "Helvetica Neue",
                "fontSize": 16,
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 12,
                "position": "absolute",
                "top": 25,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                  Object {
                    "translateX": 0,
                  },
                ],
              },
            ]
          }
        >
          Class Code
        </Text>
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Helvetica Neue",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontFamily": "Helvetica Neue",
                "fontSize": 16,
                "left": 0,
                "opacity": 1,
                "paddingHorizontal": 12,
                "position": "absolute",
                "top": 25,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                  Object {
                    "translateX": 0,
                  },
                ],
              },
            ]
          }
        >
          Class Code
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder=""
        placeholderTextColor="rgba(0, 0, 0, 0.54)"
        selectionColor="#6200ee"
        style={
          Array [
            Object {
              "flexGrow": 1,
              "fontSize": 16,
              "margin": 0,
              "minHeight": 58,
              "paddingHorizontal": 12,
              "textAlign": "left",
              "zIndex": 1,
            },
            Object {
              "minHeight": 64,
              "paddingBottom": 16,
              "paddingTop": 20,
            },
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
              "textAlignVertical": "center",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      onLayout={[Function]}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#B00020",
            "fontSize": 12,
            "opacity": 0,
            "paddingHorizontal": 12,
            "paddingVertical": 4,
            "transform": Array [],
          },
        ]
      }
    >
      Classroom codes must be exactly 4 characters long.
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "height": "10%",
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#ffa500",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "maxWidth": "60%",
          "minWidth": "60%",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "Helvetica Neue",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "color": "#000000",
                    "fontFamily": "HelveticaNeue-Medium",
                  },
                  Object {
                    "fontFamily": "HelveticaNeue-Medium",
                  },
                ],
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 24,
                  "fontWeight": "bold",
                }
              }
            >
              Connect
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
